src/plugins/optimize-graph-plugin.js

```diff
@@ -176,15 +176,31 @@ class ChunkGraph {
 				let rep = typeof newUrl === 'function' ? newUrl(url, fn, quote) : newUrl;
 				if (rep === false) return 'null';
 				if (typeof rep === 'string') {
-					if (publicPath) rep = posix.join(publicPath, rep);
-					rep = JSON.stringify(rep);
+					rep = publicPath ? toImport(publicPath, rep, true) : JSON.stringify(rep);
 				}
 				return `${fn}(${rep})`;
 			}
 		});
 	}
 }
 
+/**
+ * @param {string} publicPath
+ * @param {string} filename
+ * @param {boolean} [toJSON]
+ */
+function toImport(publicPath, filename, toJSON) {
+	let value = posix.join(publicPath, filename);
+
+	if (/^(https?:)?\/\//.test(publicPath)) {
```

Comment:
There's (amazingly) no `src/utils` directory yet, but extracting this logic (same as `toImport` inside optimize) is a good candidate to start it off.

Reply:
lol - yeah I've waffled a lot on a utils directory/module. This plugin probably warrants moving into its own repo at some point, and I didn't want to couple it to WMR's codebase as much as possible.

(agreed though)

Reply:
Yup – figured that was the case knowing you 😆 Left it this way with mental note to come back if/when commons are extracted.

Reply:
Ah yeah - I had completely ignored `publicPath` being a URL here, and assumed it was a `pathname`. We could switch from using `posix.join()` to using something like:

```js
function relative(filename, publicPath) {
  let href = new URL(publicPath, 'https://f.example.com');
  let url = new URL(filename, href);
  if (href.origin === 'https://f.example.com') return url.pathname + url.search;
  return url.href;
}
```

Reply:
You're suggesting this helper to take the place of the full `toImport` util? 
It _looks_ like it should be doing the same thing but I might be overlooking something.

Reply:
Could make it always JSON-stringify, I don't think there are any cases left where it needs to skip that.

Reply:
oh, wow I literally hadn't scrolled down to see that you'd already written this haha. I actually prefer your explicit protocol check over my origin weirdness.

FWIW `posix.join()` would still mangle the protocol in `publicPath` since it collapses adjacent slashes.

```js
/**
 * @param {string} publicPath
 * @param {string} filename
 */
function toImport(publicPath, filename, toJSON) {
	if (/^(https?:)?\/\//.test(publicPath)) {
        return JSON.stringify(new URL(filename, publicPath).href);
	} else {
		return JSON.stringify(posix.join(publicPath, filename));
	}
}
```

Reply:
The `toJSON` is a parameter because of this TODO, which needs a regular string to figure out if/what addl processing needs to be done : https://github.com/preactjs/wmr/pull/213/files#diff-6b6c2718597d2a9b4a4334e7c5f6208ffe00946ac37ded1a170cdaa6cfc7dd06R401

Your most recent snippet fails the `//example.com/asset.jpg` case. That's why my `toImport` version deals with that `isFull` and `root` juggling act. Using `new URL` fails when the base does not include a fully qualified protocol.

Reply:
Ah right I always forget Node doesn't support protocol-relative URLs (boo). So maybe this combo:

```js
/**
 * @param {string} publicPath
 * @param {string} filename
 */
function toImport(publicPath, filename) {
	const FAKE = 'https://a.a';
	const href = new URL(publicPath, FAKE);
	const url = new URL(filename, href);
	return JSON.stringify(href.origin === FAKE ? url.pathname + url.search : url.href);
}
```

<img width="347" alt="Screen Shot 2020-12-03 at 3 57 36 PM" src="https://user-images.githubusercontent.com/105127/101087488-484ca280-3580-11eb-9538-e05d629ec7cf.png">

Reply:
Treating `//example.com` the same as `https://example.com` is a mistake.

While we 100% hope/want people to be using HTTPS, that's not what they entered. The `publicPath` should be respected as is. Use of `//example.com` may be intentional so that websites always resolve using the _current_ protocol. 

(This happens.)

Reply:
With the last commit, these are the `toImport` outputs using your test values. They are expected:

```js
toImport('/', './foo/bar.js'); 
//=> '/foo/bar.js'

toImport('//example.com', './foo/bar.js')
//=> '//example.com/foo/bar.js'

toImport('//example.com', '../foo/bar.js')
//=> '//example.com/foo/bar.js'

toImport('//example.com', 'foo/bar.js')
//=> '//example.com/foo/bar.js'

toImport('http://example.com', 'foo/bar.js')
//=> 'http://example.com/foo/bar.js'

toImport('https://example.com', 'foo/bar.js')
//=> 'https://example.com/foo/bar.js'
```